File: Interview/DS/Sorting/Counting_sort/counting_sort.java
import java.util.Arrays;

public class CountingSort {
    public static void main(String[] args) {
        int[] arr = {5, 1, 4, 2, 8};
        int[] sortedArray = countingSort(arr, 8);
        System.out.println("Sorted array: " + Arrays.toString(sortedArray));
    }

    /**
     * The main counting sort method that takes an array and the maximum value in the array as its arguments.
     * 
     * @param arr the array to be sorted
     * @param maxValue the maximum value in the array
     * @return the sorted array
     */
    public static int[] countingSort(int[] arr, int maxValue) {
        int[] count = new int[maxValue + 1];
        int[] sortedArray = new int[arr.length];

        // store count of each element
        for (int i = 0; i < arr.length; i++) {
            count[arr[i]]++;
        }

        // change count[i] so that count[i] now contains actual
        // position of this element in output array
        for (int i = 1; i <= maxValue; i++) {
            count[i] += count[i - 1];
        }

        // build the output array
        for (int i = arr.length - 1; i >= 0; i--) {
            sortedArray[count[arr[i]] - 1] = arr[i];
            count[arr[i]]--;
        }

        return sortedArray;
    }
}

File: Interview/Algorithms/Sorting/Counting_sort/counting_sort.py
def count_sort(arr):
    n = len(arr)
    m = max(arr) + 1
    count = [0] * m              
    output = [0] * n
 
    for i in range(0, n):
        count[arr[i]] += 1
 
    for i in range(1, m):
        count[i] += count[i - 1]
 
    for i in range(n - 1, -1, -1):
        output[count[arr[i]] - 1] = arr[i]
        count[arr[i]] -= 1
 
    return output

# Example usage of count_sort
arr = [4, 2, 2, 8, 3, 3, 1]
sorted_array = count_sort(arr)
print("Sorted array is:", sorted_array)